### ISSUE TITLE
Dropout Tutorial purely on Tensorflow

### ISSUE THREAD

**Issue by rowedenny** [tensorflow/cleverhans#972]
I am trying to implement a model purely on Tensorflow following the tutorial but get stuck with dropout. Unfortunately, all models in the tutorial with dropout are implemented by Keras.

When I implement with Tensorflow, I define a Placeholder within the class function `fprop` of the model.
`def fprop(self, x, **kwargs):
        del kwargs

        with tf.variable_scope(self.scope, reuse=tf.AUTO_REUSE):
            self.keep_prob = tf.placeholder(tf.float32, name='keep_prob')
            print(self.keep_prob)`

Then we I train the model, I feed value to the placeholder by parameter `feed`, the code are the same with the tutorial,
`train(sess, loss, x_train, y_train, args=train_params, feed={model.keep_prob: 0.5}, rng=rng)`

But it raises the error that
`You must feed a value for placeholder tensor 'target_model_1/keep_prob' with dtype float`
so I try to print the name of the placeholder to check.

What I find wired is that, though my scope name is 'target_model', but the name of this placeholder is named with "target_model_1:keep_prob". I assume this is the reason why the error raised.

I am wondering if anyone can provide a tutorial purely on Tensorflow, showing how to add dropout within cleverhans?

**Comment by goodfeli** [comment]
Here's a model with dropout that I used recently.


class Model_dropout(NoRefModel):
  """
  Copied from cifar10_model.Model_dropout
  """

  def __init__(self, scales=3, filters=256, scope='deep_classify',
               nb_classes=10, logit_norm=True,
               orig_logit_norm=False,
               logit_norm_eps=1e-8, logit_scale=1.,
               gcn=False,
               normalize_hid=False,
               input_shape=(None, 28, 28, 1), **kwargs):
    self.logit_norm = logit_norm
    self.orig_logit_norm = orig_logit_norm
    self.logit_norm_eps = logit_norm_eps
    self.input_shape = input_shape
    self.scales = scales
    self.filters = filters
    self.logit_scale = logit_scale
    self.gcn = gcn
    self.normalize_hid = normalize_hid
    assert not normalize_hid  # Removed this feature
    super(type(self), self).__init__(nb_classes=nb_classes,
                                     needs_dummy_fprop=True,
                                     scope=scope, **kwargs)

  def fprop(self, x,
            mnist_model_hid_keep_prob=1.,
            **kwargs):
    # with tf.device('/CPU:0'):
    #  x = tf.Print(x, [tf.reduce_max(tf.abs(x))], message="x")
    scales = self.scales
    filters = self.filters
    scope = self.scope
    act = tf.nn.leaky_relu
    logit_norm_eps = self.logit_norm_eps
    logit_scale = self.logit_scale
    gcn = self.gcn
    conv_args = dict(kernel_size=3, activation=act,
                     padding='same', data_format='channels_last')
    if gcn:
      x = x - tf.reduce_mean(x, axis=(1, 2, 3), keep_dims=True)
      x = x * tf.rsqrt(tf.reduce_mean(tf.square(x),
                                      axis=(1, 2, 3), keep_dims=True))
    with tf.variable_scope(scope, reuse=tf.AUTO_REUSE):
      y = tf.layers.conv2d(x, filters, **conv_args)
      for scale in range(scales):
        y = tf.layers.conv2d(y, filters << scale, **conv_args)
        y = tf.layers.conv2d(y, filters << (scale + 1), **conv_args)
        y = downscale2d(y, 2)
        if mnist_model_hid_keep_prob != 1.:
          y = tf.nn.dropout(y, keep_prob=mnist_model_hid_keep_prob)
        # with tf.device('/CPU:0'):
        #    y = tf.Print(y, [tf.reduce_max(tf.abs(y))], message='y' +
str(scale))
      assert self.nb_classes == 10
      y = tf.layers.conv2d(y, self.nb_classes, 3, name="class_conv")
      logits = tf.reduce_mean(y, [1, 2])
      # with tf.device('/CPU:0'):
      #      logits = tf.Print(logits,
[tf.reduce_max(tf.abs(logits))], message="logits", summarize=10)
      if self.orig_logit_norm:
        logits *= logit_scale * \
            tf.rsqrt(tf.reduce_mean(
                tf.square(logits)) + logit_norm_eps)
        assert not self.logit_norm
      elif self.logit_norm:
        logits *= logit_scale * \
            tf.rsqrt(tf.reduce_mean(tf.square(logits),
                                    axis=1, keep_dims=True) + logit_norm_eps)
      # with tf.device('/CPU:0'):
      #      logits = tf.Print(logits,
[tf.reduce_max(tf.abs(logits))], message="normalized logits",
summarize=10)
      return {'logits': logits}

  def make_input_placeholder(self):
    return tf.placeholder(tf.float32, self.input_shape, 'x')

  def make_label_placeholder(self):
    return tf.placeholder(tf.float32, (None, self.nb_classes))
